Git part 2 lesson (#52)
CodeToCommunicate/CoCoLessons: 08_git_part_2.md

<img align="center" style="margin:0 0 20px 0" width="100%" height="100%" src="https://raw.githubusercontent.com/CodeToCommunicate/CoCoLessons/main/media/coco-banner.jpg">

# Working with branches | Trabajando con ramas

## What we know | Qué sabemos

> - We know how to work with the `main` (master) branch | Sabemos como trabajar

Reviewer: ```suggestion
- We know how to work with the `main` (master) branch. | Sabemos como trabajar
```

  en la rama `main` (master).

- Changes are save it in the repository as _commits_ | Los cambios se guardan
  en el repositorio como _commits_ (`git add` y `git commit`).
  <img src="data/git/git-staging-area.svg" alt="" style="height: 65%;">

- We know how to create _remote_ repositories in GitHub and link them with the
  local repositories usng `git push` and `git pull` | Sabemos cómo crear
  repositorios _remote_ en GitHub y vincularlos con los repositorios locales
  usando `git push` y `git pull`.
  <img src="data/git/github-repo-after-first-push.svg" alt="" style="height: 65%;">

**_git_ is at its best when we start working with branches within our repository |
El máximo esplendor de _git_ se obtiene cuando empezamos a trabajar con ramas
dentro de nuestro repositorio.**

## What is a branch and how do we work with it? | ¿Qué es una rama y cómo trabajamos con ella?

Branches are one of the core concepts in _git_.

> And there's an endless amount of things you can do with them.

Reviewer: ```suggestion
and there's an endless amount of things you can do with them.
```

One of the most important functions of _git_ is the control of branches to

> achieve a better development of our project.

Reviewer: No entendi que quieres decir con esto

> The _branches_ help us have multiple versions of an project organized.

Reviewer: ```suggestion
The _branches_ help us have multiple versions of a project in an organized format.
```

> In general, a new branch is a bifurcation of the state of project that creates

Reviewer: ```suggestion
In general, a new branch is a bifurcation of the current project that creates
```

Entonces, creamos una nueva rama (`fix-bug`) para probar diferentes formas de
solucionarlo.
<img src="data/git/git-repo-3.svg" alt="" style="height: 25%;">

In this new branch, we made several commits to fix bug.
At the same time, we made other commits in the `main` branch.

---

En esta nueva rama, hacimos varios commits para corregir el error.
Al mismo tiempo, hicimos otros commits en la rama `main`.
<img src="data/git/git-repo-5.svg" alt="" style="height: 25%;">

Now that we could fix the bug in the `fix-bug` branch, we want to add these
commits to the `main` branch using the `git merge` command.

---

Ahora que pudimos arreglar el error en la rama `fix-bug`, queremos agregar
estos commits a la rama `main` usando el comando `git merge`:
<img src="data/git/git-repo-6.svg" alt="" style="height: 25%;">

### Advantages of using branches | Ventajas de usar ramas

Using _git branches_ has multiple advantages.
However, we want to emphasize the following two:

- It is possible to develop new features/changes in our project without
  hindering the development in the main branch.

- It is possible to create different development branches that can converge
  in the same repository.

>   **For example, a stable branch (`main`), a test branch, and an `fix-bug`

Reviewer: ```suggestion
  **For example, a stable branch (`main`), a test branch, and a `fix-bug`
```

---

Escribir mal el nombre de una rama o simplemente cambiar de opinión sobre su
nombre después de crearla es fácil.
Es por eso que _git_ nos permit cambiar el nombre de una rama local facilmente.
Si deseamos cambiar el nombre de su rama actual, puede usar el siguiente comando:

```
$ git branch -m <nuevo-nombre>
```

En caso de que deseeamos cambiar el nombre de una rama local diferente en la
que actualmente NO está localizados, debemos proporcionar el nombre antiguo y
el nuevo:

```
$ git branch -m <nombre-antiguo> <nombre-nuevo>
```

Estos comandos, de nuevo, se utilizan para trabajar con ramas locales.

## How to merge a branch? | Como fusionar ramas

### In a local machine | En una máquina local

Finally, there comes a point where we have made many modifications to a
`<new-branch-name>` branch.
So we want to add all these modifications to the `main` branch.
For that, there is the `git merge` command.

We will attach the `<new-branch-name>` branch to the `main` branch.
First, we have to place ourselves in the `main` branch:

```
$ git swich main
```

Then, we merge the branches with:

```
git merge <new-branch-name>
```

> Now, all the modification did in <new-branch-name> are in the main branch.

Reviewer: ```suggestion
Now, all the modification we made in <new-branch-name> are in the main branch.
```

Using the command `git log`, we can see the commits.

---

Finalmente, llega un punto en el que hemos hecho muchas modificaciones a la
rama `<new-branch-name>`.
Entonces queremos fucionar estos cambios a la rama `main`.
Para eso, existe el comando `git merge`.

Fusionaremos la rama `<new-branch-name>` a la rama `main`.
Primero, tenemos que situarnos en la rama `main`:

```
$ git swich main
```

Luego, fusionamos las ramas con:

```
git merge <new-branch-name>
```

Ahora, todas las modificaciones realizadas en `<new-branch-name>` están en la
rama `main`.
Usando el comando `git log`, podemos ver los commits.

### Using GitHub | Usando GitHub

Another way tho marge a branch to `main` is using the _Pull Request_ in the
GitHub interface.
This workflow has some advantages:

- It is more visual
- You can ask review or opinion of your modifications before merge it to main.

> - It has some tool that help you to not mess up.

Reviewer: ```suggestion
- It has some tools that help you to not make a mistake.
```

Este flujo de trabajo tiene algunas ventajas:

- Es más visual.
- Puede solicitar una revisión u opinión de sus modificaciones antes de
  fusionarlas con la principal.
- Tiene algunas herramientas que te ayuda a no equivocarte.

## How to delete a branch | Como eliminar una rama

Not all branches are meant to live forever.
In fact, most branches in any repository will be short-lived.
So if we want to do a little housecleaning, the command to delete a local
branch is:

---

No todas las ramas están destinadas a vivir para siempre.
De hecho, la mayoría de las ramas en cualquier repositorio serán de corta
duración.
Entonces, si se desea hacer una pequeña limpieza, el comando par eliminar una
rama de nuesto repositorio local es:

```
$ git branch -d <branch-name>
```

## How to track branches | Como traquear ramas

By default, local and remote branches have nothing to do with each other.
They are stored and managed as independent objects in _git_.
But in real life, of course, local and remote branches often do have a
relationship with each other.
**Such a relationship can be modelled in _git_: one branch (typically a local
one) can "track" another one (typically remote)**.

> Therefor to track a remote branch:

Reviewer: ```suggestion
Therefore, to track a remote branch:
```
